### PR TITLE
[2.x] Fixed day of year defaulting for round up parser (#4627)

### LINKED ISSUE #4285: Bug in JavaDateFormatter when using DOY
**Is your feature request related to a problem? Please describe.**
Elasticsearch Version
7.10.2

Installed Plugins
No response

Java Version
1.8

OS Version
CentOS

**Problem Description**
Customer is experiencing ES v7.10 regression in handling custom date formats. Elasticsearch is not returning correct results when a date range query is specified with an inclusive upper bound on a custom date format. This is a regression from Elasticsearch 6.x, where the underlying Joda time library handled our format by default (i.e., it was not "custom"). We updated mappings for new indices in Elasticsearch 7, but Elasticsearch claims the same format that successfully indexes is not valid only when an upper bound is specified. This issue only affects indices on Elasticsearch 7 clusters; indices migrated from Elasticsearch 6 are not affected. I will attach steps to reproduce to this ticket. 

We store and query many of our date-time fields with a specific day-of-year time, e.g. 2022-111T00:00:00.000. Elasticsearch 6 automatically handled this format as a date and our mappings did not need additional customization. The format is embedded in many tools and procedures, so updating all of them and retraining our teams is not feasible. The built-ins for ES7 date fields are all just slightly off, e.g. they require the timezone to be specified but tools downstream would need to change to include that in their searches to. 

Here are more details: 

For indices we are creating on ES7, we are specifying this in the mappings for indices that will store these dates: "format": "strict_date_optional_time||yyyy-DDD'T'HH:mm:ss.SSS" 

For example, if a document exists with a property "timestamp"="2022-199T14:08:30.294" with it's mapping set to "format": "strict_date_optional_time||yyyy-DDD'T'HH:mm:ss.SSS": 

timestamp:[2022-199T14:08:30.294 TO 2022-200T14:48:05.538] returned some objects but not one in particular. 

Changing query string to be exclusive on the upper end returns the result successfully: 

timestamp:[2022-199T14:08:30.294 TO 2022-200T14:48:05.538} (exclusive on the upper end) 

Keeping the inclusive upper bound and switching to month-day also works 

timestamp:[2022-199T14:08:30.294 TO 2022-07-19T14:48:05.538] 

When translating the inclusive upper bound, custom formatted date to a DSL query, an error is returned that indicates the "format" in the mapping is not being respected: 

"root_cause" : 
[
 {
 "type" : "parse_exception",
 "reason" : "failed to parse date field [2022-199T14:08:30.294] with format [strict_date_optional_time||yyyy-DDD'T'HH:mm:ss.SSS]: [failed to parse date field [2022-199T14:08:30.294] with format [strict_date_optional_time||yyyy-DDD'T'HH:mm:ss.SSS]]" 
} 
         ],"

*** Here is GitHub issue for Elasticsearch our customer filed on the public Elasticsearch GitHub. it is this issue https://github.com/elastic/elasticsearch/issues/89096 ***

### PR DESCRIPTION
**Pull request by Vishalks**
Signed-off-by: Vishal Sarda <vsarda@amazon.com>
(cherry picked from commit a17692d0124e808b9539384e9107a2ee4a0ebe22)

### Description
This change fixes the conflict gets when trying to round up input date ranges, specifically for the format "yyyy-DDD'T'HH:mm:ss.SSS". For this input format, we default day of year to 1 else default month of year and day of month to 1

### Issues Resolved
#4285 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by andrross — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -45,8 +45,13 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
-- [Bug]: Fixed invalid location of JDK dependency for arm64 architecture([#4613](https://github.com/opensearch-project/OpenSearch/pull/4613))
+- [Segment Replication] Ignore lock file when testing cleanupAndPreserveLatestCommitPoint ([#4544](https://github.com/opensearch-project/OpenSearch/pull/4544))
+- Updated jackson to 2.13.4 and snakeyml to 1.32 ([#4556](https://github.com/opensearch-project/OpenSearch/pull/4556))
+- Fixed day of year defaulting for round up parser ([#4627](https://github.com/opensearch-project/OpenSearch/pull/4627))
 - Fixed the SnapshotsInProgress error during index deletion ([#4570](https://github.com/opensearch-project/OpenSearch/pull/4570))
+- [Segment Replication] Adding check to make sure checkpoint is not processed when a shard's shard routing is primary ([#4630](https://github.com/opensearch-project/OpenSearch/pull/4630))
+- [Bug]: Fixed invalid location of JDK dependency for arm64 architecture([#4613](https://github.com/opensearch-project/OpenSearch/pull/4613))
+- [Bug]: Alias filter lost after rollover ([#4499](https://github.com/opensearch-project/OpenSearch/pull/4499))
```
Comment (tlfeng) [inline CHANGELOG.md:54]: Line 48, 52, 54 have not included in `2.x` branch yet, could you remove those lines?

Comment (Vishalks) [reply CHANGELOG.md]: Removed

File: CHANGELOG.md [hunk 2]
```diff
@@ -45,8 +45,10 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Getting security exception due to access denied 'java.lang.RuntimePermission' 'accessDeclaredMembers' when trying to get snapshot with S3 IRSA ([#4469](https://github.com/opensearch-project/OpenSearch/pull/4469))
 - Fixed flaky test `ResourceAwareTasksTests.testTaskIdPersistsInThreadContext` ([#4484](https://github.com/opensearch-project/OpenSearch/pull/4484))
 - Fixed the ignore_malformed setting to also ignore objects ([#4494](https://github.com/opensearch-project/OpenSearch/pull/4494))
-- [Bug]: Fixed invalid location of JDK dependency for arm64 architecture([#4613](https://github.com/opensearch-project/OpenSearch/pull/4613))
+- Updated jackson to 2.13.4 and snakeyml to 1.32 ([#4556](https://github.com/opensearch-project/OpenSearch/pull/4556))
+- Fixed day of year defaulting for round up parser ([#4627](https://github.com/opensearch-project/OpenSearch/pull/4627))
 - Fixed the SnapshotsInProgress error during index deletion ([#4570](https://github.com/opensearch-project/OpenSearch/pull/4570))
+- [Bug]: Fixed invalid location of JDK dependency for arm64 architecture([#4613](https://github.com/opensearch-project/OpenSearch/pull/4613))
```
Comment (dreamer-89) [inline CHANGELOG.md:51]: Thanks @Vishalks for this change. 

Why are we changing the existing changelog entries ?

Comment (tlfeng) [reply CHANGELOG.md:51]: I believe that's a result of resolving the conflict between main and 2.x branches.
I don't think it's a big problem here, he only added one line to this part `Updated jackson to 2.13.4 and snakeyml to 1.32`, which existed in this location in `main` branch, but existed in a below line in this file of `2.x` branch.

Comment (tlfeng) [reply CHANGELOG.md:51]: Of course the changelog in `2.x` branch is very confusing, anyway it will be refactored to be more clear soon.

Comment (Vishalks) [reply CHANGELOG.md:51]: I need to change my way of resolving merge conflicts which is leading to changing the unrelated entries.

**Pull request merged**